File: go/telemetry_test.go
package telemetry

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

const endpoint = "https://netcrunch.example/api/rest/1/sensors/example@1/update"

func newTest(t *testing.T, options Options) *Telemetry {
	t.Helper()
	if options.Endpoint == "" {
		options.Endpoint = endpoint
	}
	stats, err := New(options)
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	return stats
}

// server returns a test server whose URL embeds a recognisable secret, so a leak
// into an error message is visible.
func server(t *testing.T, handler func(w http.ResponseWriter, count int)) (*httptest.Server, *[]map[string]any) {
	srv, received, _ := serverWithHeaders(t, handler)
	return srv, received
}

func serverWithHeaders(
	t *testing.T,
	handler func(w http.ResponseWriter, count int),
) (*httptest.Server, *[]map[string]any, *[]http.Header) {
	t.Helper()
	var mu sync.Mutex
	received := []map[string]any{}
	headers := []http.Header{}

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, _ := io.ReadAll(r.Body)
		var parsed map[string]any
		_ = json.Unmarshal(body, &parsed)

		mu.Lock()
		received = append(received, parsed)
		headers = append(headers, r.Header.Clone())
		count := len(received)
		mu.Unlock()

		handler(w, count)
	}))
	t.Cleanup(srv.Close)
	return srv, &received, &headers
}

func secretURL(srv *httptest.Server) string {
	return srv.URL + "/api/rest/1/sensors/SENSORSECRET@1/update"
}

// -- construction -----------------------------------------------------------

func TestNewRejectsBadEndpoint(t *testing.T) {
	for _, bad := range []string{"", "not-a-url", "ftp://host/x"} {
		if _, err := New(Options{Endpoint: bad}); err == nil {
			t.Errorf("New(%q) = nil error, want one", bad)
		}
	}
}

func TestRetainMustOutlastFlushInterval(t *testing.T) {
	// A 60s flush against a 1 minute retain would let values expire between sends.
	_, err := New(Options{Endpoint: endpoint, FlushInterval: time.Minute, Retain: time.Minute})
	if err == nil {
		t.Fatal("expected an error when retain does not exceed the flush interval")
	}
}

// -- counter handles --------------------------------------------------------

func TestSamePathResolvesToSameHandle(t *testing.T) {
	stats := newTest(t, Options{})

	first := stats.MustCounter("Queue", "Depth", "inbound")
	second := stats.MustCounter("Queue", "Depth", "inbound")
	if first != second {
		t.Error("resolving the same path twice returned different handles")
	}
	if first == stats.MustCounter("Queue", "Depth", "outbound") {
		t.Error("different instances shared a handle")
	}
}

func TestMaxAndMinMoveOneWay(t *testing.T) {
	stats := newTest(t, Options{})

	peak := stats.MustCounter("SNMP", "Peak ms", "")
	peak.Max(120)
	peak.Max(90)
	peak.Max(200)
	if got := peak.Value(); got != 200 {
		t.Errorf("Max: got %v, want 200", got)
	}

	floor := stats.MustCounter("SNMP", "Floor ms", "")
	floor.Set(100)
	floor.Min(150)
	floor.Min(40)
	floor.Min(70)
	if got := floor.Value(); got != 40 {
		t.Errorf("Min: got %v, want 40", got)
	}
}

// -- lifetime-bound aggregates ---------------------------------------------

func TestSelfCountClosesIdempotently(t *testing.T) {
	// defer and an explicit Close can both fire on the same value. A double
	// decrement drives the count negative, which drifts away from every threshold
	// rather than towards one, so nothing would ever report it.
	stats := newTest(t, Options{})
	handle := stats.MustCounter("Pool", "Leases Active", "")

	lease := stats.SelfCount("Pool", "Leases Active", "")
	if got := handle.Value(); got != 1 {
		t.Fatalf("after SelfCount: got %v, want 1", got)
	}

	lease.Close()
	lease.Close()
	lease.Close()
	if got := handle.Value(); got != 0 {
		t.Errorf("after repeated Close: got %v, want 0", got)
	}
}

func TestPartCountWithdrawsItsContribution(t *testing.T) {
	stats := newTest(t, Options{})
	handle := stats.MustCounter("Cache", "Entries", "")
	handle.Set(1000)

	part := stats.PartCount("Cache", "Entries", "")
	for _, value := range []float64{5, 9, 3} {
		if err := part.Set(value); err != nil {
			t.Fatalf("Set(%v): %v", value, err)
		}
	}
	if got := handle.Value(); got != 1003 {
		t.Errorf("after sets: got %v, want 1003", got)
	}

	part.Close()
	if got := handle.Value(); got != 1000 {
		t.Errorf("after Close: got %v, want 1000 (the pre-existing value must survive)", got)
	}
	if err := part.Set(1); err == nil {
		t.Error("Set after Close should report an error rather than silently reopening")
	}
}

func TestCategoryMovesBetweenInstances(t *testing.T) {
	stats := newTest(t, Options{})
	phase := stats.Category("Workers", "By Phase")

	phase.Set("parsing")
	if got := stats.MustCounter("Workers", "By Phase", "parsing").Value(); got != 1 {
		t.Fatalf("parsing: got %v, want 1", got)
	}

	phase.Set("writing")
	if got := stats.MustCounter("Workers", "By Phase", "parsing").Value(); got != 0 {
		t.Errorf("parsing after move: got %v, want 0", got)
	}
	if got := stats.MustCounter("Workers", "By Phase", "writing").Value(); got != 1 {
		t.Errorf("writing: got %v, want 1", got)
	}

	phase.Close()
	if got := stats.MustCounter("Workers", "By Phase", "writing").Value(); got != 0 {
		t.Errorf("writing after Close: got %v, want 0", got)
	}
}

// -- payload ----------------------------------------------------------------

func TestEmptyCollectionsAreOmitted(t *testing.T) {
	stats := newTest(t, Options{})

	body, err := stats.BuildPayload(time.Now())
	if err != nil {
		t.Fatalf("BuildPayload: %v", err)
	}
	if got := string(body); got != `{"retain":5,"remove":1440}` {
		t.Errorf("empty payload: got %s", got)
	}
}

func TestTimestampMessageCarriesNoMilliseconds(t *testing.T) {
	stats := newTest(t, Options{})
	observed := time.Date(2026, 8, 10, 9, 14, 22, 512_000_000, time.UTC)
	if err := stats.Timestamp("Sync", "Age s", "Last Sync", observed); err != nil {
		t.Fatalf("Timestamp: %v", err)
	}

	body, _ := stats.BuildPayload(time.Date(2026, 8, 10, 9, 15, 54, 0, time.UTC))
	var parsed map[string]any
	_ = json.Unmarshal(body, &parsed)

	statuses := parsed["statuses"].(map[string]any)
	message := statuses["Last Sync"].(map[string]any)["message"]
	if message != "2026-08-10T09:14:22Z" {
		t.Errorf("message: got %v", message)
	}

	// Milliseconds are dropped from the message but still count towards the age:
	// 91.488s rounds to 91, not the 92 a whole-second observation would give.
	age := parsed["counters"].([]any)[0].(map[string]any)["value"]
	if age != float64(91) {
		t.Errorf("age: got %v, want 91", age)
	}
}

// -- sending ----------------------------------------------------------------

func TestFlushClearsEventsButKeepsCountersAndStatuses(t *testing.T) {
	srv, received := server(t, func(w http.ResponseWriter, _ int) { w.WriteHeader(http.StatusOK) })
	stats := newTest(t, Options{Endpoint: srv.URL})

	stats.MustCounter("Job", "Items", "").Set(7)
	_ = stats.Status("Job", "OK")
	_ = stats.Event("started")

	if err := stats.Flush(context.Background()); err != nil {
		t.Fatalf("first flush: %v", err)
	}
	if err := stats.Flush(context.Background()); err != nil {
		t.Fatalf("second flush: %v", err)
	}

	if len(*received) != 2 {
		t.Fatalf("requests: got %d, want 2", len(*received))
	}
	if _, present := (*received)[0]["events"]; !present {
		t.Error("first payload should carry the event")
	}
	if _, present := (*received)[1]["events"]; present {
		t.Error("second payload should not repeat the event")
	}
	if _, present := (*received)[1]["counters"]; !present {
		t.Error("counters must keep being reported")
	}
}

func TestTokenIsSentAsBearer(t *testing.T) {
	srv, _, headers := serverWithHeaders(t, func(w http.ResponseWriter, _ int) { w.WriteHeader(http.StatusOK) })

	stats := newTest(t, Options{Endpoint: srv.URL, Token: "TOKENSECRET"})
	_ = stats.Status("Job", "OK")
	if err := stats.Flush(context.Background()); err != nil {
		t.Fatalf("Flush: %v", err)
	}

	if got := (*headers)[0].Get("Authorization"); got != "Bearer TOKENSECRET" {
		t.Errorf("Authorization = %q, want %q", got, "Bearer TOKENSECRET")
	}
}

func TestNoTokenMeansNoAuthorizationHeader(t *testing.T) {
	srv, _, headers := serverWithHeaders(t, func(w http.ResponseWriter, _ int) { w.WriteHeader(http.StatusOK) })

	stats := newTest(t, Options{Endpoint: srv.URL})
	_ = stats.Status("Job", "OK")
	if err := stats.Flush(context.Background()); err != nil {
		t.Fatalf("Flush: %v", err)
	}

	if got := (*headers)[0].Get("Authorization"); got != "" {
		t.Errorf("Authorization = %q, want it absent", got)
	}
}

func TestErrorsNeverCarryTheEndpoint(t *testing.T) {
	srv, _ := server(t, func(w http.ResponseWriter, _ int) { w.WriteHeader(http.StatusUnauthorized) })
	stats := newTest(t, Options{Endpoint: secretURL(srv)})
	_ = stats.Status("Job", "OK")

	err := stats.Flush(context.Background())
	if err == nil {
		t.Fatal("expected an error")
	}

	var sendErr *Error
	if !errors.As(err, &sendErr) || sendErr.StatusCode != http.StatusUnauthorized {
		t.Fatalf("got %#v, want *Error with status 401", err)
	}
	// The URL is currently the credential; it must not reach a log.
	if strings.Contains(err.Error(), "SENSORSECRET") || strings.Contains(err.Error(), "127.0.0.1") {
		t.Errorf("error leaked the endpoint: %v", err)
	}
}

func TestRetryPolicy(t *testing.T) {
	rejected, rejectedRequests := server(t, func(w http.ResponseWriter, _ int) { w.WriteHeader(http.StatusBadRequest) })
	stats := newTest(t, Options{Endpoint: rejected.URL, MaxRetries: 3})
	_ = stats.Status("Job", "OK")

	if err := stats.Flush(context.Background()); err == nil {
		t.Fatal("expected a 400 to fail")
	}
	if len(*rejectedRequests) != 1 {
		t.Errorf("a rejected request must not be repeated: got %d attempts", len(*rejectedRequests))
	}

	flaky, flakyRequests := server(t, func(w http.ResponseWriter, count int) {
		if count == 1 {
			w.WriteHeader(http.StatusServiceUnavailable)
			return
		}
		w.WriteHeader(http.StatusOK)
	})
	retrying := newTest(t, Options{Endpoint: flaky.URL, MaxRetries: 1})
	_ = retrying.Status("Job", "OK")

	if err := retrying.Flush(context.Background()); err != nil {
		t.Fatalf("a 503 should be retried and then succeed: %v", err)
	}
	if len(*flakyRequests) != 2 {
		t.Errorf("attempts: got %d, want 2", len(*flakyRequests))
	}
}

func TestNothingStagedMeansNothingSent(t *testing.T) {
	srv, received := server(t, func(w http.ResponseWriter, _ int) { w.WriteHeader(http.StatusOK) })
	stats := newTest(t, Options{Endpoint: srv.URL})

	if err := stats.Flush(context.Background()); err != nil {
		t.Fatalf("Flush: %v", err)
	}
	if len(*received) != 0 {
		t.Errorf("requests: got %d, want 0", len(*received))
	}
}

// TestConcurrentUse is about the race detector rather than its assertions.
func TestConcurrentUse(t *testing.T) {
	var served atomic.Int64
	srv, _ := server(t, func(w http.ResponseWriter, _ int) {
		served.Add(1)
		w.WriteHeader(http.StatusOK)
	})
	stats := newTest(t, Options{Endpoint: srv.URL})

	handle := stats.MustCounter("Load", "Operations", "")

	var wg sync.WaitGroup
	for worker := 0; worker < 8; worker++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for i := 0; i < 200; i++ {
				handle.Inc()
				lease := stats.SelfCount("Load", "In Flight", "")
				_ = stats.Status("Worker", "OK")
				lease.Close()
			}
			_ = stats.Flush(context.Background())
		}()
	}
	wg.Wait()

	if got := handle.Value(); got != 1600 {
		t.Errorf("operations: got %v, want 1600", got)
	}
	if got := stats.MustCounter("Load", "In Flight", "").Value(); got != 0 {
		t.Errorf("in flight: got %v, want 0", got)
	}
}

File: js/test/telemetry.test.js
import assert from "node:assert/strict";
import http from "node:http";
import test from "node:test";

import { Telemetry, TelemetryError } from "../src/index.js";

const ENDPOINT = "https://netcrunch.example/api/rest/1/sensors/example@1/update";

const connect = (options = {}) =>
  new Telemetry({ endpoint: ENDPOINT, detectLeaks: false, ...options });

/** Starts a throwaway server whose URL embeds a recognisable secret. */
function startServer(handler) {
  return new Promise((resolve) => {
    const requests = [];
    const server = http.createServer((req, res) => {
      const chunks = [];
      req.on("data", (chunk) => chunks.push(chunk));
      req.on("end", () => {
        requests.push(JSON.parse(Buffer.concat(chunks).toString("utf8") || "{}"));
        handler(res, requests.length);
      });
    });
    server.listen(0, "127.0.0.1", () => {
      const { port } = server.address();
      resolve({
        server,
        requests,
        url: `http://127.0.0.1:${port}/api/rest/1/sensors/SENSORSECRET@1/update`,
        close: () => new Promise((done) => server.close(done)),
      });
    });
  });
}

// -- construction -----------------------------------------------------------

test("endpoint must be an absolute http or https URL", () => {
  assert.throws(() => new Telemetry({}), /endpoint is required/);
  assert.throws(() => new Telemetry({ endpoint: "not-a-url" }), /absolute http or https/);
  assert.throws(() => new Telemetry({ endpoint: "ftp://host/x" }), /absolute http or https/);
});

test("retain must outlast the flush interval", () => {
  // 60s flush against a 1 minute retain would let values expire between sends.
  assert.throws(
    () => new Telemetry({ endpoint: ENDPOINT, flushSeconds: 60, retainMinutes: 1 }),
    /must exceed flushSeconds/
  );
});

// -- counter handles --------------------------------------------------------

test("the same path always resolves to the same handle", () => {
  const stats = connect();
  const a = stats.counter("Queue", "Depth", "inbound");
  const b = stats.counter("Queue", "Depth", "inbound");
  assert.equal(a, b);
  assert.notEqual(a, stats.counter("Queue", "Depth", "outbound"));
});

test("max and min only move the value one way", () => {
  const stats = connect();
  const peak = stats.counter("SNMP", "Peak ms");
  peak.max(120).max(90).max(200);
  assert.equal(peak.value, 200);

  const floor = stats.counter("SNMP", "Floor ms").set(100);
  floor.min(150).min(40).min(70);
  assert.equal(floor.value, 40);
});

// -- lifetime-bound aggregates ---------------------------------------------

test("selfCount holds +1 until disposed", () => {
  const stats = connect();
  const handle = stats.counter("Pool", "Leases Active");

  const first = stats.selfCount("Pool", "Leases Active");
  const second = stats.selfCount("Pool", "Leases Active");
  assert.equal(handle.value, 2);

  first.dispose();
  assert.equal(handle.value, 1);
  second.dispose();
  assert.equal(handle.value, 0);
});

test("disposing twice does not decrement twice", () => {
  // `using` and an explicit dispose() can both fire on the same object. A double
  // decrement would corrupt the count far more quietly than a missing one.
  const stats = connect();
  const handle = stats.counter("Pool", "Leases Active");
  const lease = stats.selfCount("Pool", "Leases Active");

  lease.dispose();
  lease.dispose();
  lease[Symbol.dispose]();

  assert.equal(handle.value, 0);
  assert.equal(lease.disposed, true);
});

test("partCount withdraws exactly what it contributed", () => {
  const stats = connect();
  const handle = stats.counter("Cache", "Entries").set(1000);
  const part = stats.partCount("Cache", "Entries");

  part.set(5);
  part.set(9);
  part.set(3);
  assert.equal(handle.value, 1003);
  assert.equal(part.contribution, 3);

  part.dispose();
  assert.equal(handle.value, 1000);
});

test("category moves its +1 between instances", () => {
  const stats = connect();
  const phase = stats.category("Workers", "By Phase");

  phase.set("parsing");
  assert.equal(stats.counter("Workers", "By Phase", "parsing").value, 1);

  phase.set("writing");
  assert.equal(stats.counter("Workers", "By Phase", "parsing").value, 0);
  assert.equal(stats.counter("Workers", "By Phase", "writing").value, 1);

  phase.set("writing"); // no-op
  assert.equal(stats.counter("Workers", "By Phase", "writing").value, 1);

  phase.dispose();
  assert.equal(stats.counter("Workers", "By Phase", "writing").value, 0);
});

// -- payload ----------------------------------------------------------------

test("empty collections are omitted rather than sent empty", () => {
  const stats = connect();
  assert.deepStrictEqual(stats.buildPayload(), { retain: 5, remove: 1440 });

  stats.status("Phase", "running");
  const payload = stats.buildPayload();
  assert.ok(payload.statuses);
  assert.equal(payload.counters, undefined);
  assert.equal(payload.events, undefined);
});

test("timestamp messages carry no milliseconds", () => {
  const stats = connect();
  stats.timestamp("Sync", "Age s", "Last Sync", { observedAt: new Date("2026-08-10T09:14:22.512Z") });

  const payload = stats.buildPayload({ snapshotAt: new Date("2026-08-10T09:15:54.000Z") });
  // Milliseconds are dropped from the message but still count towards the age:
  // 91.488s rounds to 91, not to the 92 a whole-second observation would give.
  assert.equal(payload.statuses["Last Sync"].message, "2026-08-10T09:14:22Z");
  assert.equal(payload.counters[0].value, 91);
});

// -- sending ----------------------------------------------------------------

test("a successful flush clears events but keeps counters and statuses", async (t) => {
  const target = await startServer((res) => res.writeHead(200).end("{}"));
  t.after(target.close);

  const stats = connect({ endpoint: target.url });
  stats.counter("Job", "Items").set(7);
  stats.status("Job", "OK");
  stats.event("started");

  await stats.flush();
  await stats.flush();

  assert.equal(target.requests.length, 2);
  assert.deepStrictEqual(target.requests[0].events, [{ message: "started" }]);
  assert.equal(target.requests[1].events, undefined);
  assert.equal(target.requests[1].counters[0].value, 7);
  assert.deepStrictEqual(target.requests[1].statuses, { Job: { value: "OK" } });
});

test("errors never carry the endpoint", async (t) => {
  const target = await startServer((res) => res.writeHead(401).end("nope"));
  t.after(target.close);

  const stats = connect({ endpoint: target.url });
  stats.status("Job", "OK");

  await assert.rejects(stats.flush(), (error) => {
    assert.ok(error instanceof TelemetryError);
    assert.equal(error.statusCode, 401);
    // The URL is currently the credential — it must not reach a log or a stack.
    assert.doesNotMatch(error.message, /SENSORSECRET|127\.0\.0\.1/);
    assert.doesNotMatch(error.stack ?? "", /SENSORSECRET/);
    return true;
  });
});

test("4xx is not retried; 5xx is", async (t) => {
  const rejected = await startServer((res) => res.writeHead(400).end());
  t.after(rejected.close);

  const stats = connect({ endpoint: rejected.url, maxRetries: 3 });
  stats.status("Job", "OK");
  await assert.rejects(stats.flush(), /HTTP 400/);
  assert.equal(rejected.requests.length, 1, "a rejected request must not be repeated");

  const flaky = await startServer((res, count) =>
    count === 1 ? res.writeHead(503).end() : res.writeHead(200).end("{}")
  );
  t.after(flaky.close);

  const retrying = connect({ endpoint: flaky.url, maxRetries: 1 });
  retrying.status("Job", "OK");
  await retrying.flush();
  assert.equal(flaky.requests.length, 2, "a 503 must be retried");
});

test("concurrent flushes share one request", async (t) => {
  const target = await startServer((res) => setTimeout(() => res.writeHead(200).end("{}"), 50));
  t.after(target.close);

  const stats = connect({ endpoint: target.url });
  stats.counter("Job", "Items").set(1);

  await Promise.all([stats.flush(), stats.flush(), stats.flush()]);
  assert.equal(target.requests.length, 1);
});

test("nothing staged means nothing sent", async (t) => {
  const target = await startServer((res) => res.writeHead(200).end("{}"));
  t.after(target.close);

  await connect({ endpoint: target.url }).flush();
  assert.equal(target.requests.length, 0);
});
